Feature/#70 autocomplete in cli

frankschmitt opened the pull request:
Add a command for generating auto-completion scripts for Bash and Zsh. Implements #70

frankschmitt committed:
fix typo in README

frankschmitt committed:
#70: bump clikt dependency to v3.5.0

frankschmitt committed:
#70: provide default Properties() object (fixes NPE)

frankschmitt committed:
:memo: Update CLI configuration documentation

frankschmitt committed:
#70: add generate-completion option + command (WIP)

frankschmitt committed:
#70: implement generate-completion command

frankschmitt committed:
Merge remote-tracking branch 'origin/feature/#70-autocomplete-in-cli' into feature/#70-autocomplete-in-cli

frankschmitt committed:
:memo: Update CLI configuration documentation

frankschmitt committed:
#70: provide gradle task to generate Bash and Zsh completion scripts

frankschmitt committed:
#70: unit tests for generating completion scripts (WIP)

frankschmitt committed:
#70: write meaningful unit tests for auto-completion generation

We use a custom version of execAppInSubprocess() that directly writes the output to a file. This avoids a messed-up console when generating the scripts.

frankschmitt committed:
#70: add empty logback config for unit tests

Because logback clutters STDOUT, we have to use a custom logback config that simply doesn't log anything.

frankschmitt committed:
Merge branch '4sh:dev' into feature/#70-autocomplete-in-cli

Lysoun commented on buildSrc/src/main/kotlin/Versions.kt:
```diff
@@ -10,7 +10,8 @@ object Versions {
     const val mockk = "1.9.3"
-    const val clikt = "2.3.0"
+    //const val clikt = "2.3.0"
+    const val clikt = "3.5.0"
```
You can remove that comment

frankschmitt replied:
Done.

Lysoun commented on modules/cli/build.gradle.kts:
```diff
@@ -17,7 +17,8 @@ dependencies {
 
-    implementation("com.github.ajalt:clikt:${Versions.clikt}")
+    implementation("com.github.ajalt.clikt:clikt:${Versions.clikt}")
+    //implementation("com.github.ajalt:clikt:${Versions.clikt}")
```
You could remove the comment

frankschmitt replied:
Done.

Lysoun commented on modules/cli/src/test/kotlin/datamaintain/cli/app/AppTest.kt:
```diff
@@ -201,5 +205,18 @@ internal class AppTest : BaseCliTest() {
+            fun setupLogger() {
+                logger.addAppender(testAppender)
+                testAppender.start()
+            }
```
This test class does nothing, doesn't it?

frankschmitt replied:
Doh! Seems I forgot to remove that class after moving tests to a separate sub-package, sorry. I've removed it.

Lysoun commented on modules/cli/src/main/kotlin/datamaintain/cli/completion/RebuildAutoCompletionScripts.kt:
```diff
@@ -0,0 +1,50 @@
+    listOf("bash", "zsh").forEach {
+        val result = execAppInSubprocess(listOf("--generate-completion", it),
+                                 outputDir + "/" + it + "-autocomplete.sh")
+        //File(outputDir + "/" + it + "-autocomplete.sh").writeText(result.output)
```
This commented line could be deleted

frankschmitt replied:
Done.

Lysoun reviewed (changes requested):
That looks very promising! Did you try using the autocompletion? I sourced the `zsh-automplete.sh` file and tried using the CLI but couldn't find any completion. Maybe I used it wrongly? You can build the CLI using the following command in the CLI module
```gradle installDist```
and then you will find it in `modules/cli/build/install/clin/bin`

In my opinion, the scripts should not be committed in the repository. They should be included and sourced in our docker packages and in the CLI file that is uploading on every release but if it sounds difficult to you, I can do it. 

I'd prefer the option to build the CLI not to be available in `App` to avoid displaying it in the help and making it a valid option for our users to call. Could you change the implementation and make it read a global variable, that will have to be set during our releases?

frankschmitt committed:
#70: cleanup after code review

frankschmitt committed:
#70: cleanup after code review, part II

frankschmitt commented:
> That looks very promising! Did you try using the autocompletion? I sourced the `zsh-automplete.sh` file and tried using the CLI but couldn't find any completion. Maybe I used it wrongly? 

I guess that's because currently, the autocompletion uses the (very non-specific) name "app". See my comment at https://github.com/4sh/datamaintain/issues/70#issuecomment-1278163455 .

After generating the zsh script and sourcing it, auto-completion for the "app" command should work like this

<pre><font color="#EF2929"><b>➜  </b></font><font color="#34E2E2"><b>datamaintain</b></font> <font color="#729FCF"><b>git:(</b></font><font color="#EF2929"><b>feature/#70-autocomplete-in-cli</b></font><font color="#729FCF"><b>) </b></font><font color="#FCE94F"><b>✗</b></font> source bla.sh
<font color="#8AE234"><b>➜  </b></font><font color="#34E2E2"><b>datamaintain</b></font> <font color="#729FCF"><b>git:(</b></font><font color="#EF2929"><b>feature/#70-autocomplete-in-cli</b></font><font color="#729FCF"><b>) </b></font><font color="#FCE94F"><b>✗</b></font> app
generate-completion      list                     mark-script-as-executed  update-db              
</pre>

> You can build the CLI using the following command in the CLI module `gradle installDist` and then you will find it in `modules/cli/build/install/clin/bin`

Nice! I was wondering how to run the CLI, actually. Thanks!

> In my opinion, the scripts should not be committed in the repository. They should be included and sourced in our docker packages and in the CLI file that is uploading on every release but if it sounds difficult to you, I can do it.

I'll give it at shot.

> I'd prefer the option to build the CLI not to be available in `App` to avoid displaying it in the help and making it a valid option for our users to call. Could you change the implementation and make it read a global variable, that will have to be set during our releases?

Sure (I'm not yet sure whether I need to exclude the class somehow from compilation altogether or whether wrapping the call that registers the command is sufficient, but I'll figure it out).

frankschmitt committed:
#70: only provide generateCompletion if global flag is set

frankschmitt committed:
:memo: Update CLI configuration documentation

frankschmitt committed:
#70: only generate completion with gradle task

We now only provide generation of completion scripts via the gradle task rebuildCompletion. To achieve this, the following changes were necessary:
- removal of completionCommand() from App
- introduction of an "alternative" AppWithCompletion

Also, some cleanup was overdue:
- removal of obsolete *flag* --generate-completion (we only use the *task* generate-completion)

frankschmitt committed:
Merge remote-tracking branch 'origin/feature/#70-autocomplete-in-cli' into feature/#70-autocomplete-in-cli

frankschmitt commented:
@Lysoun I've adapted the code; in principle, it should now work as required:
- the task generate-completion is no longer available in App (and therefore, end users won't see it)
- the gradle task for generating the completion scripts is still available (so devs can create it)
- the completion scripts for bash and zsh were removed; I've added an empty .gitkeep file to the directory docs/auto-completion because it's required for the gradle task
- the docker build scripts now call the gradle task for generating auto completion

However, a couple of things still need to be done / clarified:
- app name for auto-completion. This is currently still "app"; I could create a sub-class of App with a more sensible name (e.g. Datamaintain) and use that for generating the scripts. However, then we would have mismatch between the name used for auto-completion and the "real" app name.
  I'm not sure how we should handle this. One possible approach would be to rename App.kt to Datamaintain.kt (or similar) - do you think that's sensible?
- regarding the Docker images: I'm not sure how that's supposed to work. Should the auto-completion scripts be deployed in the Docker container? Or do you expect users to create an alias for running datamaintain via docker
  (something like alias dm="docker run --it --rm datamaintain:<release>-mongo") and use the auto-completion on the Docker host?
- regarding the CLI file: how is this packaged? As far as I can see, cli / cli.bat is auto-generated by Gradle. Should the auto-completion scripts be bundled insided cli.tar?

frankschmitt committed:
#70: code cleanup

Lysoun commented:
> However, a couple of things still need to be done / clarified:
> 
> * app name for auto-completion. This is currently still "app"; I could create a sub-class of App with a more sensible name (e.g. Datamaintain) and use that for generating the scripts. However, then we would have mismatch between the name used for auto-completion and the "real" app name.
>   I'm not sure how we should handle this. One possible approach would be to rename App.kt to Datamaintain.kt (or similar) - do you think that's sensible?

I think you could just rename the class `App` without renaming the file, though it's named quite badly; it's not clear that the CLI's main is in that file.

> * regarding the Docker images: I'm not sure how that's supposed to work. Should the auto-completion scripts be deployed in the Docker container? Or do you expect users to create an alias for running datamaintain via docker
>   (something like alias dm="docker run --it --rm datamaintain:-mongo") and use the auto-completion on the Docker host?

Actually we already have docker packages that run Datamaintain so I was not thinking of an alias. I was thinking of deploying the auto-completion scripts and put them in the `.bashrc` of the dockers packages so that you can take advantage from it when you are inside the container. All the Dockerfiles used to build our images are in `modules/cli/Dockerfile`.

> * regarding the CLI file: how is this packaged? As far as I can see, cli / cli.bat is auto-generated by Gradle. Should the auto-completion scripts be bundled insided cli.tar?

I would find it nice, with a little README to explain how to enable autocompletion (just give the set of commands)

frankschmitt commented:
Sounds good, I‘ll
- rename App to something more unique
- adapt the Dockerfiles / build process
- adapt the build process for cli.tar

frankschmitt committed:
#70: rename App to Datamaintain (for more sensible auto-completion)

frankschmitt committed:
#70: deploy auto-completion scripts with release + Docker

- include README and completion scripts in release.tar
- include completion scripts in Docker images, and source them
- fix test failures (use correct path for renamed main file)

frankschmitt commented:
@Lysoun I've implemented the changes. Auto-completion now "listens" to datamaintain, and the scripts are included in the release and the Docker images. Looking forward to your review :-)

Lysoun commented on docs/auto-completion/README.md:
````diff
@@ -0,0 +1,16 @@
+
+# BUILD
+Run 
+```./gradlew rebuildCompletion```
````
```suggestion
```./gradlew rebuildAutocompletion```
```

Lysoun commented on modules/cli/Dockerfile/mongo/4/Dockerfile:
```diff
@@ -30,4 +30,8 @@ RUN mkdir /code/datamaintain \
+RUN echo "source /code/datamaintain/bash-autocomplete.sh" >> ~/.bashrc \
+    && echo "source /code/datamaintain/zsh-autocomplete.sh" >> ~/.zshrc
+
 ENTRYPOINT ["/code/datamaintain/bin/cli"]
```
```suggestion
ENTRYPOINT ["datamaintain"]
```

Lysoun commented on modules/cli/Dockerfile/mongo/5/Dockerfile:
```diff
@@ -27,4 +27,8 @@ RUN mkdir /code/datamaintain \
+RUN echo "source /code/datamaintain/bash-autocomplete.sh" >> ~/.bashrc \
+    && echo "source /code/datamaintain/zsh-autocomplete.sh" >> ~/.zshrc
+
 ENTRYPOINT ["/code/datamaintain/bin/cli"]
```
```suggestion
ENTRYPOINT ["datamaintain"]
```

Lysoun commented on modules/cli/Dockerfile/mongo/4/Dockerfile:
```diff
@@ -30,4 +30,8 @@ RUN mkdir /code/datamaintain \
+# Setup autocompletion
+RUN echo "source /code/datamaintain/bash-autocomplete.sh" >> ~/.bashrc \
+    && echo "source /code/datamaintain/zsh-autocomplete.sh" >> ~/.zshrc
+
```
```suggestion
RUN alias datamaintain=/code/datamaintain/bin/cli
```

Lysoun commented on modules/cli/Dockerfile/mongo/5/Dockerfile:
```diff
@@ -27,4 +27,8 @@ RUN mkdir /code/datamaintain \
+# Setup autocompletion
+RUN echo "source /code/datamaintain/bash-autocomplete.sh" >> ~/.bashrc \
+    && echo "source /code/datamaintain/zsh-autocomplete.sh" >> ~/.zshrc
+
```
```suggestion
RUN alias datamaintain=/code/datamaintain/bin/cli
```

Lysoun commented on docs/auto-completion/README.md:
````diff
@@ -0,0 +1,16 @@
+# USAGE
+Source the script for your shell, e.g. for Bash:
+
+```source bash-autocomplete.sh```
````
```suggestion
```source bash-autocomplete.sh```
Define an alias to the Datamaintain CLI:
```alias datamaintain=/your/path/cli```
```

Lysoun reviewed (approved):
Thank you so much for your work, I'm approving your PR!

Lysoun committed:
Update docs/auto-completion/README.md

Lysoun committed:
Update docs/auto-completion/README.md

Lysoun committed:
Add alias to datamaintain in docker packages

Lysoun commented:
I'm gonna merge it in branch 2.1 after fixing some conflicts
